### ISSUE TITLE
GC warnings on large allocations

### ISSUE THREAD

**Issue by yxhuvud** [crystal-lang/crystal#11274]
When allocating or resizing large arrays or hash tables the GC will output the following warning:
```
GC Warning: Repeated allocation of very large block (appr. size 134221824):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 1073745920):
	May lead to memory leak and poor performance
```

This feels quite ugly to me and it would be nice if Crystal would allow big containers without emitting warnings.

In #4441 which also discussed the issue it was suggested to set the `GC_LARGE_ALLOC_WARN_INTERVAL` variable. This however feels quite ugly to me as it involves action from the user and it is not very discoverable.

While looking for some other information I ran into https://github.com/ivmai/bdwgc/blob/master/doc/debugging.md#warning-messages-about-needing-to-allocate-blacklisted-blocks , which suggests to use `GC_malloc_ignore_off_page` for dealing with the issue. That is a `GC_malloc` replacement that does not warn on big allocations. So what does using it mean? Well, any memory allocated by it will be collected if not pointed to by a pointer to somewhere in the first 512 bytes of the memory. Or in other words, any pointer to INSIDE the internals of the memory is unsafe. This can be contrasted to with `GC_malloc_atomic` which means internal pointers pointing OUT from the memory won't be collected. There is also a `GC_malloc_atomic_ignore_off_page` for when the allocator cares neither about pointers pointing in or out of the inside of allocations which could be interesting too, but not for this particular discussion as the containers in question definitely need to keep what they contain alive.

One notable lack of the API is that there is only a `malloc`, not a `realloc` so `realloc` needs to be implemented, and it will involve an extra write to the size of the original allocation. This means it is slightly slower than the original `realloc` and should only be used when necessary. The choice between warning and performance is a bit unfun. 

So is `GC_malloc_ignore_off_page` safe for hashes or arrays? Well, in the case of hash I'd say that it definitely is, as I can't think of any reasonable way to get a pointer into the internal allocations. In the case of Array it is slightly more controversial as arrays have `to_unsafe` and people could do whatever with that pointer. That already doesn't work if people resize the array (as `GC_realloc` invoke `GC_free` on the old allocation) but it is at least a bit feasible that someone somewhere could initialize a big array and then keep the allocation alive by a pointer to the inside. 

Anyhow, I implemented a proof of concept in https://gist.github.com/yxhuvud/1d3f03a11fa0cc45f4d4574529068ee6. 

Current implementation:
```
GC Warning: Repeated allocation of very large block (appr. size 201330688):
	May lead to memory leak and poor performance
GC Warning: Repeated allocation of very large block (appr. size 1610616832):
	May lead to memory leak and poor performance
   1  28.33M ( 35.29ns) (±16.27%)    128B/op               fastest
  10   5.14M (194.39ns) (± 9.90%)    448B/op          5.51× slower
 100 538.58k (  1.86µs) (± 9.18%)  4.14kB/op         52.61× slower
  1k  47.05k ( 21.25µs) (± 4.98%)  32.5kB/op        602.19× slower
 10k   3.84k (260.39µs) (± 8.75%)   513kB/op       7377.66× slower
100k 346.09  (  2.89ms) (± 3.08%)  4.75MB/op      81867.22× slower
  1M  24.55  ( 40.73ms) (± 8.45%)  39.8MB/op    1153921.94× slower
 10M 805.41m (  1.24s ) (± 0.99%)   640MB/op   35178860.75× slower
100M  73.12m ( 13.68s ) (± 0.00%)   5.0GB/op  387509098.57× slower
```
This monkeypatch:
```
   1  27.56M ( 36.28ns) (±15.40%)    128B/op               fastest
  10   5.29M (188.98ns) (± 9.23%)    448B/op          5.21× slower
 100 538.74k (  1.86µs) (± 8.41%)  4.14kB/op         51.16× slower
  1k  47.28k ( 21.15µs) (± 9.25%)  32.5kB/op        582.95× slower
 10k   3.83k (260.80µs) (± 8.18%)   513kB/op       7187.47× slower
100k 337.42  (  2.96ms) (± 2.66%)  4.75MB/op      81676.93× slower
  1M  22.86  ( 43.75ms) (± 5.73%)  39.8MB/op    1205716.64× slower
 10M 783.16m (  1.28s ) (± 0.90%)   640MB/op   35190337.11× slower
100M  70.33m ( 14.22s ) (± 0.00%)   5.0GB/op  391868489.88× slower
```
So as you can see the difference is small (and often a lot less than the variance between runs), though the last one is a bit more noticeable. As far as I've been able to determine, that difference is due to the lack of `realloc`.

Thoughts? Is this a path worth pursuing? If so, do we want it only for `Hash` or also for `Array` allocations?

**Comment by asterite** [comment]
I think we should just silence/ignore all GC warnings and that's it.

**Comment by yxhuvud** [comment]
Hmm, any pointers to where that could be done? I'm finding no initialization in gc/boehm.cr .

**Comment by asterite** [comment]
https://github.com/ivmai/bdwgc/blob/dee02534adbdb1ea03e7ba2772bdee22e744df04/include/gc.h#L1399

We use that and set a function that does nothing. Let me know if that works!

**Comment by straight-shoota** [comment]
Can we just set `GC_LARGE_ALLOC_WARN_INTERVAL` to a large number? Only if the env var is not already defined. This would allow to override the default configuration at runtime.

**Comment by straight-shoota** [comment]
I don't think the information is completely useless. It still tells a thing about what's going on. So it has *some* value. But I agree that it's not much.

Looking from a sysop perspective, I'd like to be able to enable such verbose output when I notice something wonky is going on with an application. And I shouldn't have to rebuild the binary for that, or worse, even modify the source code.

The environment variable works great for that. There could certainly be more useful output, of course.

Maybe a combination of both could be a good solution: We add a custom warn proc in stdlib, and it can be configured via environment variable. By default it should be disabled.

**Comment by asterite** [comment]
Do you know of any language where a GC outputs warnings, or where you can configure these? I don't.

These warnings are just a boehm GC thing. We don't need to expose any of this to the users. If we later change to another GC that doesn't have warnings, what do we do?

I really think we should just silence all warnings, unless we can prove these warnings are useful.

**Comment by straight-shoota** [comment]
Yeah, it's probably fine to just disable them.

**Comment by rdp** [comment]
Could GC_malloc_ignore_off_page be used for large arrays and large hashes (like the warning message kind of implies) and GC_malloc_atomic_ignore_off_page  for large Strings?  Or not?  Or leave that for a separate discussion?

**Comment by rdp** [comment]
I'll admit I saw this the other day `GC Warning: Repeated allocation of very large block (appr. size 266240):` and wondered "what is allocating something so large? why'? I guess this would help me answer that, thanks for the tip.

**Comment by asterite** [comment]
Right, but there's no way to know what that memory of chunk is related to...
